[java] Parse error with yield lambda inside switch

**Affects PMD Version:** 

PMD 7.20.0 (fa478ec141c154d34aac5389f0d5add0f28ecfea, 2025-12-30T14:37:52Z)

**Description:**

Parsing code fails while it is valid Java.

**Exception Stacktrace:**

```
[WARN] This analysis could be faster, please consider using Incremental Analysis: https://docs.pmd-code.org/pmd-doc-7.20.0/pmd_userdocs_incremental_analysis.html
../tools/build/Lambda.java	-	ParseException: Parse exception in file '../tools/build/Lambda.java' at line 6, column 27: Encountered <IDENTIFIER: "a">.
Was expecting one of:
    "instanceof" ...
    ")" ...
    "," ...
    "=" ...
    "<" ...
    "?" ...
    "==" ...
    "<=" ...
    ">=" ...
    "!=" ...
    "||" ...
    "&&" ...
    "++" ...
    "--" ...
    "+" ...
    "-" ...
    "*" ...
    "/" ...
    "&" ...
    "|" ...
    "^" ...
    "%" ...
    "+=" ...
    "-=" ...
    "*=" ...
    "/=" ...
    "&=" ...
    "|=" ...
    "^=" ...
    "%=" ...
    "<<=" ...
    ">>=" ...
    ">>>=" ...
    ">" ...
[INFO] An error occurred while executing PMD.
```

**Code Sample demonstrating the issue:**

```java
public class Lambda {

   public static void main( String[] args ) {
      foo(switch ( args[0] ) {
         default:
            yield (Lambda a, Lambda b) -> "hello";
      });
   }

   private static void foo(java.util.function.BiFunction<Lambda, Lambda, String> f)  {}        
}
```

**Rules**

```xml
<?xml version="1.0"?>

<ruleset name="Custom Rules"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">

    <description>
        Dead code
    </description>

    <rule ref="category/java/bestpractices.xml/UnusedPrivateMethod">
        <properties>
            <property name="ignoredAnnotations" value="jakarta.annotation.PostConstruct,jakarta.annotation.PreDestroy" />
        </properties>
    </rule>

</ruleset>
```

**Steps to reproduce:**

Run `pmd check -f text -R deadcode.xml -d testdir`. Running on Java 25.

Having a single parameter in the lambda does _not_ trigger the parse error.

The code sample does not compile with java:

```
Lambda.java:4: error: no suitable method found for println(<none>)
      System.out.println(switch ( args[0] ) {
                ^
    method PrintStream.println() is not applicable
      (actual and formal argument lists differ in length)
    method PrintStream.println(boolean) is not applicable
      (argument mismatch; boolean is not a functional interface)
    method PrintStream.println(char) is not applicable
      (argument mismatch; char is not a functional interface)
    method PrintStream.println(int) is not applicable
      (argument mismatch; int is not a functional interface)
    method PrintStream.println(long) is not applicable
      (argument mismatch; long is not a functional interface)
    method PrintStream.println(float) is not applicable
      (argument mismatch; float is not a functional interface)
    method PrintStream.println(double) is not applicable
      (argument mismatch; double is not a functional interface)
    method PrintStream.println(char[]) is not applicable
      (argument mismatch; Array is not a functional interface)
    method PrintStream.println(String) is not applicable
      (argument mismatch; String is not a functional interface)
    method PrintStream.println(Object) is not applicable
      (argument mismatch; Object is not a functional interface)
1 error
error: compilation failed
```

you are yielding a lambda from the switch expression, which is not a suitable argument for println...

Can you make an example, that compiles?

--- a/pmd-java/src/main/javacc/Java.jjt
+++ b/pmd-java/src/main/javacc/Java.jjt
@@ -501,18 +501,19 @@ class JavaParserImpl {
                int lookahead = offset + 1;
                int balance = 1;
                JavaccToken t;
+               boolean hasComma = false;
                while ((t = getToken(lookahead)) != null && balance > 0) {
                    switch (t.kind) {
                        case LPAREN: balance++; break;
                        case RPAREN: balance--; break;
-                       case COMMA: if (balance == 1) return false; // a method call, eg yield(1, 2);
+                       case COMMA: if (balance == 1) { hasComma = true; } break;
                    }
                    lookahead++;
                }
                // lambda:       yield () -> {};
                // method call:  yield ();
                return t != null
-                   && (lookahead != offset + 2 // ie ()
+                   && (!hasComma && lookahead != offset + 2 // ie ()
                    || t.kind == LAMBDA);
            default:
                return false;

--- a/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ParserCornersTest.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ParserCornersTest.java
@@ -196,6 +196,27 @@ class ParserCornersTest extends BaseJavaTreeDumpTest {
     }
 
     @Test
+    void testYieldStmtWithLambda() {
+        java15.parse(
+            "public class Lambda {\n"
+            + "    interface Func {\n"
+            + "        String apply(Lambda a, Lambda b);\n"
+            + "    }\n"
+            + "\n"
+            + "   public static void main( String[] args ) {\n"
+            + "      foo(switch ( args[0] ) {\n"
+            + "         default:\n"
+            + "            yield (Lambda a, Lambda b) -> \"hello\";\n"
+            + "      });\n"
+            + "   }\n"
+            + "\n"
+            + "   private static void foo(Func f)  {}        \n"
+            + "}"
+        );
+    }
+
+
+    @Test
     void testUnicodeIndent() {
         // https://github.com/pmd/pmd/issues/3423
         java7.parseResource("UnicodeIdentifier.java");

